File: src/sp/it/pl/util/parsing/ConverterFX.java
package sp.it.pl.util.parsing;

import org.jetbrains.annotations.NotNull;
import sp.it.pl.util.conf.Configurable;
import sp.it.pl.util.functional.Try;
import static java.util.stream.Collectors.joining;
import static sp.it.pl.util.dev.Util.log;
import static sp.it.pl.util.functional.Try.error;
import static sp.it.pl.util.functional.Try.ok;
import static sp.it.pl.util.functional.Util.stream;

/** Converter for javaFX bean convention. */
public class ConverterFX extends Converter {

    private final String DELIMITER_CONFIG_VALUE = "-";
    private final String DELIMITER_CONFIG_NAME = ":";

    @SuppressWarnings("unchecked")
    @NotNull
    @Override
    public <T> Try<T,String> ofS(@NotNull Class<T> type, @NotNull String text) {
        try {
            String[] values = text.split(DELIMITER_CONFIG_VALUE);
            Class<?> typeValue = Class.forName(values[0]);
            if (!type.isAssignableFrom(typeValue))
                throw new Exception(); // optimization, avoids next line

            T v = (T) typeValue.getConstructor().newInstance();
            Configurable c = Configurable.configsFromFxPropertiesOf(v);
            stream(values).skip(1)
                .forEach(str -> {
                    try {
                        String[] nameValue = str.split(DELIMITER_CONFIG_NAME);
                        if (nameValue.length!=2) return; // ignore
                        String name = nameValue[0], val = nameValue[1];
                        c.setField(name, val);
                    } catch (Exception e) {
                        throw new RuntimeException(e.getMessage(), e);
                    }
                });
            return ok(v);
        } catch (Exception e) {
            log(ConverterFX.class).warn("Parsing failed, class={} text={}", type, text, e);
            return error(e.getMessage());
        }
    }

    @NotNull
    @Override
    public <T> String toS(T o) {
        return o.getClass().getName() + DELIMITER_CONFIG_VALUE +
            Configurable.configsFromFxPropertiesOf(o).getFields().stream()
                .map(c -> c.getName() + DELIMITER_CONFIG_NAME + c.getValueS())
                .collect(joining(DELIMITER_CONFIG_VALUE));
    }

}

File: src/sp/it/pl/util/parsing/ConverterFX.kt
package sp.it.pl.util.parsing

import mu.KLogging
import sp.it.pl.util.conf.Configurable
import sp.it.pl.util.functional.Try
import sp.it.pl.util.functional.net

/** Converter for javaFX bean convention.  */
class ConverterFX: Converter() {

    // TODO: escape text
    private val delimiterVal = "-"
    private val delimiterName = ":"

    @Suppress("UNCHECKED_CAST")
    override fun <T> ofS(type: Class<T>, text: String): Try<T, String> {
        if (text==Parsers.DEFAULT.stringNull)
            return Try.ok(null) as Try<T, String>   // TODO: fix type-safety hazard

        return try {
            val values = text.split(delimiterVal).toTypedArray()
            val valueType = Class.forName(values[0])
            if (type.isAssignableFrom(valueType)) {
                val v = (valueType.getConstructor().newInstance() as T)!!
                val c = v.toConfigurable()
                values.asSequence()
                        .drop(1)
                        .forEach {
                            val nameValue = it.split(delimiterName).toTypedArray()
                            if (nameValue.size==2) {
                                val name = nameValue[0]
                                val value = nameValue[1]
                                c.setField(name, value)
                            } else {
                                // ignore instead of error
                            }
                        }
                Try.ok(v)
            } else {
                val message = "$valueType is not $type"
                logger.warn { message }
                Try.error(message)
            }
        } catch (e: Exception) {
            logger.warn(e) { "Parsing failed, class=$type text=$text" }
            Try.error(e.message ?: "")
        }
    }

    override fun <T: Any?> toS(o: T): String = o
            ?.net {
                val v = it as Any
                val values = v.toConfigurable().fields.joinToString(delimiterVal) { it.name+delimiterName+it.valueS }
                v::class.java.name+delimiterVal+values
            }
            ?: Parsers.DEFAULT.stringNull

    private fun Any.toConfigurable() = Configurable.configsFromFxPropertiesOf(this)

    companion object: KLogging()

}